### PR TITLE
clean up

### PR DESCRIPTION
**Pull request by paulofaria**
- Rename some APIs to be more inlined with Swift API Design Guidelines
- Remove unneeded typealiases
- Make `Query` faster and more convenient by using an underlying `Dictionary` instead of `Array`
- Make `Data` `CustomStringConvertible` and `CustomDebugStringConvertible`
- Add new line to end of all files

### CODE REVIEW

**Comment by Danappelxx** [comment]
It seems to me as if the `Query` has a lot of boilerplate where its simply calling the underlying dictionary for the properties. Is there a way to clean this up? 

Other than that LGTM

**Comment by paulofaria** [comment]
It's the same deal with `headers`, they need the `struct` so we can make the API for accessing multiple values easier to bare.

**Comment by paulofaria** [comment]
Best we handle the boilerplate then the consumers right?

**Comment by Danappelxx** [comment]
I mean stuff like `.count` - that seems like something that conforming to `Sequence` should provide.

File: Sources/Data.swift [hunk 1]
```diff
@@ -330,3 +324,49 @@ extension Data {
...
+
+        return debugDescription
+    }
+}
+
+extension Data: CustomDebugStringConvertible {
+    public var debugDescription: String {
+        return hexDescription
+    }
+}
```
Comment (scottbyrns) [inline Sources/Data.swift]: Needs new line

**Comment by paulofaria** [comment]
We need to provide `count` to conform to `Collection` haha.

File: Sources/Connection.swift [hunk 2]
```diff
@@ -1,4 +1,4 @@
-public protocol Connection: Stream {
```
Comment (scottbyrns) [inline Sources/Connection.swift:1]: Oops!

File: Sources/AsyncStream.swift [hunk 3]
```diff
@@ -1,7 +1,7 @@
 public protocol AsyncStream {
     var closed: Bool { get }
     func close() -> Bool
-    func receive(length length: Int, result: (Void throws -> Data) -> Void)
-    func send(data data: Data, result: (Void throws -> Void) -> Void)
+    func receive(upTo byteCount: Int, result: (Void throws -> Data) -> Void)
```
Comment (scottbyrns) [inline Sources/AsyncStream.swift]:
``` swift
stream.receive(upTo: 1024) { getData in
    do {
        let data = try getData()
    } catch {
        print(error)
    }
}
```

When read says "receive up to 1024 in the execution of" and I ask "1024 of what?"
Could we some how rephrase this to read as a literal statement that is self constant? Maybe adding a extensions to Int for bytes, kilobytes, megabytes, etc that each convert the Int into it's byte equivalent so that we can write it as:

``` swift
stream.receive(upTo: 1.kilobyte) { getData in
    do {
        let data = try getData()
    } catch {
        print(error)
    }
}
```

File: Sources/QueryField.swift [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    return lhs.values  == rhs.values
+}
+
+func ==<T: Equatable>(lhs: [T?], rhs: [T?]) -> Bool {
+    if lhs.count != rhs.count { return false }
+    for (l, r) in zip(lhs, rhs) {
+        if l != r { return false }
+    }
+    return true
+}
```
Comment (scottbyrns) [inline Sources/QueryField.swift]: Needs new line :P

File: Sources/AsyncStream.swift (hunk 3 above)
Comment (paulofaria) [inline Sources/AsyncStream.swift]: @tannernelson @LoganWright  hahaha yeah I proposed that in our slack discussion about this. We thought it wouldn't be needed as the function signature has `byteCount` in it.

``` swift
func receive(upTo byteCount: Int, result: (Void throws -> Data) -> Void)
```

But we also discussed how that would only be accessible by someone with an IDE. And adding `byte`, `bytes`, `kilobyte`, etc.. would make things much more readable. The argument @LoganWright used is that no one would use it. I said I would haha, and we should encourage it. Another problem @LoganWright pointed, is that having `bytes` on `Int` could clash with code from other frameworks. My view is if we worry too much about this we'll get too constrained. It's hard (maybe just impossible) to adapt to _everybody_. But also, I don't think `bytes` are 100% needed, so we could go either way. Also, those extensions could be added in another library if one really wants them.

Comment (scottbyrns) [inline Sources/AsyncStream.swift]: I'm one such that develops with out an IDE.

Perhaps we could look at using something like maxBytes as a label so that it reads more explicitly.

`receive max bytes in the execution of`

Comment (paulofaria) [inline Sources/AsyncStream.swift]: `upTo` is following Swift API Design Guidelines. From the standard library:

``` swift
/// Returns `self[startIndex..<end]`
///
/// - Complexity: O(1)
public func prefix(upTo end: Self.Index) -> Self.SubSequence
```

**Comment by dreymonde** [comment]
@paulofaria are you sure that all that `Query` boilerplate is needed? According to [swiftdoc](http://swiftdoc.org/v3.0/protocol/RangeReplaceableCollection/), there are only a bunch of required methods in `RangeReplaceableCollection`, others have default implementations, including `append`, `remove` and so on.

**Comment by paulofaria** [comment]
@dreymonde how do we know which ones are and aren't required?

**Comment by dreymonde** [comment]
@paulofaria required are marked as `required` in swiftdoc. Actually, swiftdoc is pretty self-explanatory and well-structured.

**Comment by paulofaria** [comment]
hahaha I think I might be blind. thanks.

File: Sources/AsyncStream.swift (hunk 3 above)
Comment (tanner0101) [inline Sources/AsyncStream.swift]: I like `upTo`. I also think that `.kilobytes` trick is dope.

**Comment by tanner0101** [comment]
Beautiful work @paulofaria. :+1: :shipit:

File: Sources/QueryField.swift [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    }
+}
+
+extension QueryField: Equatable {}
+
+public func ==(lhs: QueryField, rhs: QueryField) -> Bool {
+    return lhs.values  == rhs.values
+}
+
+func ==<T: Equatable>(lhs: [T?], rhs: [T?]) -> Bool {
```
Comment (loganwright) [inline Sources/QueryField.swift]: I think spaces between operator and declaration can aid readability. Particularly w/ `<` operator which might exist at some point, it gets hard to read: `func <<T>(...`

File: Sources/QueryField.swift [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        get {
+            return values[index]
+        }
+
+        set(value) {
+            values[index] = value
+        }
+    }
+
+    public subscript (bounds: Range<Int>) -> ArraySlice<String?> {
```
Comment (loganwright) [inline Sources/QueryField.swift]: extra space `script (`

File: Sources/QueryField.swift [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+extension QueryField: RangeReplaceableCollectionType, MutableCollectionType {}
+#endif
+
+extension QueryField {
+    public init() {
+        self.init([])
+    }
+
+    #if swift(>=3.0)
+    public mutating func replaceSubrange<C : Collection where C.Iterator.Element == String?>(subRange: Range<Int>, with newElements: C) {
```
Comment (loganwright) [inline Sources/QueryField.swift:21]: Should this possibly be:

```
public mutating func replace(subRange subRange: Range<Int>, with ...`
```

**Comment by loganwright** [comment]
:shipit:

File: Sources/QueryField.swift (hunk 7 above)
Comment (paulofaria) [inline Sources/QueryField.swift:21]: that's `RangeReplaceableCollection` conformance
